beacon-chain/blockchain/chain_info.go

```diff
@@ -343,8 +343,14 @@ func (s *Service) IsOptimisticForRoot(ctx context.Context, root [32]byte) (bool,
 		return true, nil
 	}
 
+	// Historical non-canonical blocks here are returned as optimistic for safety.
+	isCanonical, err := s.IsCanonical(ctx, root)
+	if err != nil {
+		return false, err
+	}
+
 	if slots.ToEpoch(ss.Slot)+1 < validatedCheckpoint.Epoch {
-		return false, nil
+		return !isCanonical, nil
```

Comment:
Something smells about this line. 
Could a root be:
1. non-canonical
2. older than validated checkpoint
then we return true?

Reply:
yes, and yes we should return true here, that was the bug.

Reply:
Ah i see it now! Thanks